Add documentation of OpenTelemetry setup with the collector (#195)
pganalyze/pganalyze-docs: collector/otel-setup/traceparent.mdx

---

> title: 'Setup traceparent query tag within your application'

Reviewer: ```suggestion
title: 'Set up traceparent query tag within your application'
```

backlink_href: /docs/collector/otel-setup
backlink_title: 'OpenTelemetry Setup'
---

In order for a tracing span created by the pganalyze collector to correctly
attach to the existing tracing on the application side, the query needs to have
a `traceparent` query tag, as defined in the [W3 Trace Context specification](https://www.w3.org/TR/trace-context/#traceparent-header).

> Query tags are key-value pairs appended to the SQL statements as query comments.

Reviewer: ```suggestion
Query tags are key-value pairs appended to SQL statements as query comments.
```

The pganalyze collector currently supports the following three formats to read
a `traceparent` tag:

* [sqlcommenter](https://google.github.io/sqlcommenter/) format: `key='value'`
* sqlcommenter format without single quotes: `key=value`
* [marginalia](https://github.com/basecamp/marginalia) format: `key:value`

## Add traceparent with Go

The easiest way to add a query tag with Go is to use a drop-in replacement for
go's `database/sql` library, provided by [sqlcommenter](https://google.github.io/sqlcommenter/go/database_sql/).

> That library provides an option `EnableTraceparent`, you can enable it to append

Reviewer: ```suggestion
That library provides an option `EnableTraceparent`: you can enable it to append
```

a traceparent query tag as a query comment. Please note that this is only
available when you're using OpenTelemetry for your application's tracing too, as

> it is going to refer the context and the trace context should be filled by the

Reviewer: ```suggestion
it is going to reference the context and the trace context should be filled by the
```

for details.

### Rails 7+

Rails 7 introduced [Active Record Query Logs](https://api.rubyonrails.org/classes/ActiveRecord/QueryLogs.html)
which automatically appends comments (query tags) to SQL queries.
You can enable this via Rails configuration in `config/application.rb` or an
initializer:

```ruby
config.active_record.query_log_tags_enabled = true
```

By default, this doesn't have a traceparent query tag, so you'll need to add it
manually. New comment tags can be defined by adding them in a `Hash` to the tags
`Array`.
You can append the traceparent tag like the following:

```ruby
config.active_record.query_log_tags = [
  :application,
  :controller,
  :action,
  :job,
  {
    traceparent: -> () {
      current_span = ::OpenTelemetry::Trace.current_span
      return if current_span == ::OpenTelemetry::Trace::Span::INVALID

      span_context = current_span.context
      trace_flag   = span_context.trace_flags.sampled? ? '01' : '00'
      format(
        '00-%<trace_id>s-%<span_id>s-%<trace_flags>.2d',
        trace_id: span_context.hex_trace_id,
        span_id: span_context.hex_span_id,
        trace_flags: trace_flag
      )
    }
  }
]
```

> With above configuration, the query will have the comment like the following:

Reviewer: ```suggestion
With above configuration, queries will have the comments like the following:
```

```sql
SELECT "articles".* FROM "articles" /*application='myblog',action='index',controller='articles',traceparent='00-953c157ad37a81d94fe83bb95cfc8add-3cf784b99bd43910-01'*/
```

### Marginalia

> If you are using Rails 6 and below, you can attach SQL comments using a

Reviewer: ```suggestion
If you are using Rails 6 and below, you can attach SQL comments using the
```

[marginalia](https://github.com/basecamp/marginalia) gem.

You can define a transparent tag in the initializer
(e.g. `config/initializers/marginalia.rb`):

```ruby
module Marginalia
  module Comment
    def self.traceparent
      current_span = ::OpenTelemetry::Trace.current_span
      return if current_span == ::OpenTelemetry::Trace::Span::INVALID

      span_context = current_span.context
      trace_flag   = span_context.trace_flags.sampled? ? '01' : '00'
      format(
        '00-%<trace_id>s-%<span_id>s-%<trace_flags>.2d',
        trace_id: span_context.hex_trace_id,
        span_id: span_context.hex_span_id,
        trace_flags: trace_flag
      )
    end
  end
end

Marginalia::Comment.components = [:application, :controller, :action, :job, :traceparent]
```

### Limitations

> If you're using the OpenTelemetry SDK, this is best to be used with `OpenTelemetry::Instrumentation::ActiveRecord`

Reviewer: ```suggestion
If you're using the OpenTelemetry SDK, this is best used with `OpenTelemetry::Instrumentation::ActiveRecord`
```

and `OpenTelemetry::Instrumentation::PG` instrumentation. Currently, the query
comment is attached by either Rails or marginalia after the ActiveRecord
instrumentation, but before the pg instrumentation.

> This makes the span created by the pganalyze collector to have a span from the

Reviewer: ```suggestion
This makes the span created by the pganalyze collector have a span from the
```

ActiveRecord instrumentation as a parent, not the pg one.
This will cause the database/query span (from pg) and EXPLAIN Plan (from the
pganalyze collector) to be a "parallel" relation, which may make it harder to

> identify which database span is associated with which EXPLAIN Plan.

Reviewer: I don't fully understand this: if I'm using the SDK, does using the above `OpenTelemetry::Instrumentation` classes avoid this problem, or is this a separate issue?

Author: Ah good point. It's a separate issue, but it's the issue happens only with these instrumentation classes. I'm going to rewrite this part a bit.

---
title: 'Set up traceparent query tag within your application'
backlink_href: /docs/collector/otel-setup
backlink_title: 'OpenTelemetry Setup'
---

In order for a tracing span created by the pganalyze collector to correctly
attach to the existing tracing on the application side, the query needs to have
a `traceparent` query tag, as defined in the [W3 Trace Context specification](https://www.w3.org/TR/trace-context/#traceparent-header).

Query tags are key-value pairs appended to SQL statements as query comments.
The pganalyze collector currently supports the following three formats to read
a `traceparent` tag:

* [sqlcommenter](https://google.github.io/sqlcommenter/) format: `key='value'`
* sqlcommenter format without single quotes: `key=value`
* [marginalia](https://github.com/basecamp/marginalia) format: `key:value`

## Add traceparent with Go

The easiest way to add a query tag with Go is to use a drop-in replacement for

> go's `database/sql` library, provided by [sqlcommenter](https://google.github.io/sqlcommenter/go/database_sql/).

Author: Good point. I aligned this with Ruby side a bit more, and made `database/sql` as a sub section of it (like how I have Rails and Marginalia in Ruby part)

That library provides an option `EnableTraceparent`: you can enable it to append
a traceparent query tag as a query comment. Please note that this is only
available when you're using OpenTelemetry for your application's tracing too, as
it is going to reference the context and the trace context should be filled by
the application side already.

## Add traceparent with Ruby

### Prerequisite

> Setup OpenTelemetry SDK with your application. Read the

Reviewer: ```suggestion
Set up OpenTelemetry SDK with your application. Read the
```

      span_context = current_span.context
      trace_flag   = span_context.trace_flags.sampled? ? '01' : '00'
      format(
        '00-%<trace_id>s-%<span_id>s-%<trace_flags>.2d',
        trace_id: span_context.hex_trace_id,
        span_id: span_context.hex_span_id,
        trace_flags: trace_flag
      )
    end
  end
end

Marginalia::Comment.components = [:application, :controller, :action, :job, :traceparent]
```

### Limitations

If you're using the OpenTelemetry SDK with Ruby, the pganalyze collector
integration is best used with `OpenTelemetry::Instrumentation::ActiveRecord` and
`OpenTelemetry::Instrumentation::PG` instrumentation.
With this setup, the EXPLAIN Plan span from the pganalyze collector will be
created as a child trace of the ActiveRecord span, making it easy to relate the
EXPLAIN Plan span to the ActiveRecord code.
However, the span from the pg instrumentation will be created as a sibling span
of the pganalyze collector, which may make it harder to relate them together.
This is due to the limitation of when we can generate a query comment, which is
currently only possible after the ActiveRecord instrumentation, but before the

> pg instrumentation.

Reviewer: I still don't quite understand what this is telling me (this is probably due to my lack of familiarity with otel at this point). Specifically, what does it mean that it's created as a sibling span, and can I do anything about that? Also, is this a consequence of using the `OpenTelemetry::Instrumentation` classes, or a separate problem. If the latter should this be two separate limitations bullet points?

No need to edit this further on my account at this point (given I have no relevant otel experience), but I wanted to note some parts that are still unclear to me.

Author: FYI: I chatted with Maciek and explained this. We agreed that this is not needed as it's kinda small detail, so I'll drop this for now.
